journalpump/rsyslog.py

```diff
@@ -135,6 +136,10 @@ def close(self):
     def send(self, message):
         if self.socket is None:
             self._connect()
+
+        if self.socket is None:
+            raise Exception("socket failed to connect")
```

Comment:
I'd argue that it is not the purpose of this function to check the success of `self._connect()`, which it seems to be doing with that check on the socker

Reply:
So this is a common structuring problem in almost all our daemons: `ProviderDaemon(someConfig) -> daemon.init() -> daemon.run() -> do_things_with_inited_resources()`

This pattern is hard to work with the typechecker because where and how things should fail is non-obvious: the answer we have is, just crash the process and let it be restarted. Which is a practical safety measure but makes it hard to modify the code because you never know if you're about to crash something somewhere else. Can self.socket be None, is that a reasonable thing (here, no, elsewhere, potentially, in both cases, nothing confirms you're being sensible or haven't missed something).

In this case, calling `self._connect()` _implies_ that self.socket is now not None ... but does nothing to assure the typechecker, or say, a programmer not looking at the rest of the code, that that's what it's actually done. So the typechecker rightly complains that self.socket can still be `Optional[socket]` after this, and in fact if someone breaks `self._connect()` in the future nothing is actually going to be able to determine it (we just crash loop).

Better - and I did start refactoring in this direction but it's a lot of changes - would be if the signature was `self._connect() -> socket.Socket`, because then the answer is explicitly "you always get a socket, or we're going to crash - don't worry about it". But it's a lot of refactoring and I don't understand the code well enough to know I wouldn't break behavior doing that.

This line could be replaced with a simple `assert self.socket is not None` to make the typechecker happy, which at runtime at least compiles down to a noop AFAIK (there's some way to do it IIRC). 

IMO the exception here is an improvement but not much of one, since if the socket actually fails to connect we kind of want to know why - DNS error (config?), did it connect previously ever etc.